From f59beb833968c77df6e7c3d47ea53aaf56736124 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Fri, 19 Sep 2025 07:37:04 +0200
Subject: [PATCH 922/954] feat(clk): support multiple clock providers via clock
 ID

Clock management now supports multiple providers by utilizing the
clock ID passed to identify the source of each clock provider.

clk_register() remains functional for legacy use, but can now be
replaced by clk_add_default_provider() for default provider
registration.

To register additional clock providers, use the new clk_add_provider()
API.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 52f85881a109d42ea6a29ae8b96d28ce1da4fda2 feat(clk): support multiple clock providers via clock ID]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I21176fa31eb89a3eb5148f8e04394d92ceb75e43
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/493934
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/clk/clk.c     | 102 +++++++++++++++++++++++++++++++++++-------
 include/drivers/clk.h |   7 +++
 2 files changed, 92 insertions(+), 17 deletions(-)

diff --git a/drivers/clk/clk.c b/drivers/clk/clk.c
index 4cbc0f70f4..60ec31de9d 100644
--- a/drivers/clk/clk.c
+++ b/drivers/clk/clk.c
@@ -11,48 +11,93 @@
 
 #include <drivers/clk.h>
 
-static const struct clk_ops *ops;
+/* Maximum number of clock providers */
+#ifdef CFG_CLK_PROVIDER_NB
+#define MAX_CLOCK_PROVIDERS CFG_CLK_PROVIDER_NB
+#else
+#define MAX_CLOCK_PROVIDERS 15U
+#endif
+
+struct clock_provider_t {
+	const struct clk_ops *ops;
+	uint16_t phandle;
+};
+
+static struct clock_providers_t {
+	struct clock_provider_t providers[MAX_CLOCK_PROVIDERS];
+	uint8_t nb;
+	uint8_t def_provider; /* Index of the default provider */
+} clk_providers;
+
+static const struct clk_ops *get_ops(unsigned long id)
+{
+	uint16_t phandle = (id & CLK_PHANDLE_MASK) >> CLK_PHANDLE_SHIFT;
+	const struct clk_ops *ops = NULL;
+	uint8_t i;
+	bool found = false;
+
+	/* If no phandle is provided, use the default ops */
+	if (phandle != 0U) {
+		/* Check if phandle is registered in the clock providers list */
+		for (i = 0U; i < clk_providers.nb; i++) {
+			if (clk_providers.providers[i].phandle == phandle) {
+				ops = clk_providers.providers[i].ops;
+				found = true;
+				break;
+			}
+		}
+	}
+
+	if (!found) {
+		ops = clk_providers.providers[clk_providers.def_provider].ops;
+	}
+
+	/* If no ops found, panic */
+	if (ops == NULL) {
+		ERROR("Clock provider not found for phandle %u\n", phandle);
+		panic();
+	}
+
+	return ops;
+}
+
 
 int clk_enable(unsigned long id)
 {
-	assert((ops != NULL) && (ops->enable != NULL));
+	const struct clk_ops *ops = get_ops(id);
 
-	return ops->enable(id);
+	return ops->enable(id & CLK_BINDING_ID_MASK);
 }
 
 void clk_disable(unsigned long id)
 {
-	assert((ops != NULL) && (ops->disable != NULL));
+	const struct clk_ops *ops = get_ops(id);
 
-	ops->disable(id);
+	ops->disable(id & CLK_BINDING_ID_MASK);
 }
 
 unsigned long clk_get_rate(unsigned long id)
 {
-	assert((ops != NULL) && (ops->get_rate != NULL));
+	const struct clk_ops *ops = get_ops(id);
 
-	return ops->get_rate(id);
+	return ops->get_rate(id & CLK_BINDING_ID_MASK);
 }
 
 int clk_get_parent(unsigned long id)
 {
-	assert((ops != NULL) && (ops->get_parent != NULL));
+	const struct clk_ops *ops = get_ops(id);
 
-	return ops->get_parent(id);
+	return ops->get_parent(id & CLK_BINDING_ID_MASK);
 }
 
 bool clk_is_enabled(unsigned long id)
 {
-	assert((ops != NULL) && (ops->is_enabled != NULL));
+	const struct clk_ops *ops = get_ops(id);
 
-	return ops->is_enabled(id);
+	return ops->is_enabled(id & CLK_BINDING_ID_MASK);
 }
 
-/*
- * Initialize the clk. The fields in the provided clk
- * ops pointer must be valid.
- */
-void clk_register(const struct clk_ops *ops_ptr)
+void clk_add_provider(uint32_t phandle, const struct clk_ops *ops_ptr)
 {
 	assert((ops_ptr != NULL) &&
 	       (ops_ptr->enable != NULL) &&
@@ -61,5 +106,28 @@ void clk_register(const struct clk_ops *ops_ptr)
 	       (ops_ptr->get_parent != NULL) &&
 	       (ops_ptr->is_enabled != NULL));
 
-	ops = ops_ptr;
+	if (clk_providers.nb >= MAX_CLOCK_PROVIDERS) {
+		ERROR("No more room for clock providers");
+		panic();
+	}
+
+	clk_providers.providers[clk_providers.nb].phandle = phandle;
+	clk_providers.providers[clk_providers.nb].ops = ops_ptr;
+	clk_providers.nb++;
+}
+
+void clk_add_default_provider(uint32_t phandle, const struct clk_ops *ops_ptr)
+{
+	clk_add_provider(phandle, ops_ptr);
+
+	clk_providers.def_provider = clk_providers.nb - 1;
+}
+
+/*
+ * Initialize the clk. The fields in the provided clk
+ * ops pointer must be valid.
+ */
+void clk_register(const struct clk_ops *ops_ptr)
+{
+	clk_add_default_provider(0, ops_ptr);
 }
diff --git a/include/drivers/clk.h b/include/drivers/clk.h
index a18f41ffca..bd45e91cd7 100644
--- a/include/drivers/clk.h
+++ b/include/drivers/clk.h
@@ -9,6 +9,10 @@
 
 #include <stdbool.h>
 
+#define CLK_PHANDLE_MASK	GENMASK_32(31, 16)
+#define CLK_BINDING_ID_MASK	GENMASK_32(15, 0)
+#define CLK_PHANDLE_SHIFT	16U
+
 struct clk_ops {
 	int (*enable)(unsigned long id);
 	void (*disable)(unsigned long id);
@@ -25,4 +29,7 @@ int clk_get_parent(unsigned long id);
 
 void clk_register(const struct clk_ops *ops);
 
+void clk_add_provider(uint32_t phandle, const struct clk_ops *ops);
+void clk_add_default_provider(uint32_t phandle, const struct clk_ops *ops);
+
 #endif /* CLK_H */
-- 
2.53.0

